internal/filter/parser.y

```diff
@@ -0,0 +1,186 @@
+	if ok && chain.op == ANY && LogicalOp(op) == ALL {
+	    // Retrieve the last pushed condition and append it to the new "And" chain instead
+	    andChain, _ := NewChain(ALL, chain.pop())
+	    andChain.add(rules...)
+
+	    chain.add(andChain)
+
+	    return chain
+	}
+
+	// If the given operator is the same as the already existsing chains operator (*chain),
+	// we don't need to create another chain of the same operator type. Avoids something
+	// like &Chain{op: ALL, &Chain{op: ALL, ...}}
+	if chain == nil || chain.op != LogicalOp(op) {
+	    newChain, err := NewChain(LogicalOp(op), rule)
+	    if err != nil {
+		// Just panic, filter.Parse will try to recover from this.
+		panic(err)
+	    }
+
+	    chain = newChain
+	}
+
+	chain.add(rules...)
+
+	return chain
+}
+%}
+
+%union {
+	expr Filter
+	text string
+}
+
+%type <expr> filter_rule
+%type <expr> filter_chain
+%type <expr> conditions_expr
+%type <expr> maybe_negated_condition_expr
+%type <expr> condition_expr
+%type <expr> exists_expr
+
+%type <text> comparison_op
+%type <text> optional_negation
+%type <text> identifier
+%type <text> logical_op
+
+%token T_EQUAL "="
+%token T_UNEQUAL "!" T_EQUAL
+%token T_LIKE "~"
+%token T_UNLIKE "!" T_LIKE
+%token T_LESS_THAN "<"
+%token T_GREATER_THAN ">"
+%token T_LESS_THAN_OR_EQUAL "<" T_EQUAL
+%token T_GREATER_THAN_OR_EQUAL ">" T_EQUAL
+%token <text> T_STRING
+%token <text> T_IDENTIFIER
+
+%type <text> T_EQUAL
+%type <text> T_UNEQUAL
+%type <text> T_LIKE
+%type <text> T_UNLIKE
+%type <text> T_LESS_THAN
+%type <text> T_GREATER_THAN
+%type <text> T_LESS_THAN_OR_EQUAL
+%type <text> T_GREATER_THAN_OR_EQUAL
+
+%type <text> "!" "&" "|"
+
+// This is just used for declaring explicit precedence and resolves shift/reduce conflicts
+// in `filter_chain` and `conditions_expr` rules.
+%nonassoc PREFER_SHIFTING_LOGICAL_OP
+
+%nonassoc T_EQUAL T_UNEQUAL T_LIKE T_UNLIKE
+%nonassoc T_LESS_THAN T_LESS_THAN_OR_EQUAL T_GREATER_THAN T_GREATER_THAN_OR_EQUAL
+
+%left "!" "&" "|"
+%left "("
+%right ")"
+
+%%
+
+filter_rule: filter_chain logical_op filter_chain
+	{
+		$$ = reduceFilter($2, $1, $3)
+		yylex.(*Lexer).rule = $$
+	}
+	| filter_chain
+	;
+
+filter_chain: conditions_expr logical_op maybe_negated_condition_expr
+	{
+		$$ = reduceFilter($2, $1, $3)
+		yylex.(*Lexer).rule = $$
+	}
+	| conditions_expr %prec PREFER_SHIFTING_LOGICAL_OP
+	;
+
+conditions_expr: maybe_negated_condition_expr logical_op maybe_negated_condition_expr
+	{
+		$$ = reduceFilter($2, $1, $3)
+		yylex.(*Lexer).rule = $$
+	}
+	| maybe_negated_condition_expr %prec PREFER_SHIFTING_LOGICAL_OP
+	;
+
+maybe_negated_condition_expr: optional_negation condition_expr
+	{
+		if $1 != "" {
+			// NewChain is only going to return an error if an invalid operator is specified, and since
+			// we explicitly provide the NONE operator, we don't expect an error to be returned.
+			$$, _ = NewChain(NONE, $2)
+		} else {
+			$$ = $2
+		}
+
+		yylex.(*Lexer).rule = $$
+	}
+
+condition_expr: "(" filter_chain ")"
```

Comment:
By resolving brackets at this point, brackets around `identifier comparison_op identifier` (i.e. `(foo)=(bar)`) are not possible.
However, I don't know if this input expression should even be accepted.

Reply:
Our IPL web parser does not accept such filter strings, so it's up to @julianbrost whether to accept such here or not:
```php
ipl\Web\Filter\ParseException : Invalid filter "(foo)=(bar)", unexpected = at pos 6: Expected logical operator
 /usr/local/src/ipl-web/src/Filter/Parser.php:561
```